lib/services/local/presence.go

```diff
@@ -657,6 +658,300 @@ func (s *PresenceService) DeleteAllRemoteClusters() error {
 	return trace.Wrap(err)
 }
 
+// jitter is a shared jitter function used to jitter internal retries.
+var jitter = utils.NewJitter()
+
+// AcquireSemaphore attempts to acquire the specified semaphore.  AcquireSemaphore will automatically handle
+// retry on contention.  If the semaphore has already reached MaxLeases, or there is too much contention,
+// a LimitExceeded error is returned.
+func (s *PresenceService) AcquireSemaphore(ctx context.Context, req services.AcquireSemaphoreParams) (*services.SemaphoreLease, error) {
+	// this combination of backoff parameters leads to worst-case total time spent
+	// in backoff between 1200ms and 2400ms depending on jitter.  tests are in
+	// place to verify that this is sufficient to resolve a 20-lease contention
+	// event, which is worse than should ever occur in practice.
+	const baseBackoff = time.Millisecond * 400
+	const acquireAttempts int64 = 4
+
+	if err := req.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if req.Expires.Before(s.Clock().Now().UTC()) {
+		return nil, trace.BadParameter("cannot acquire expired semaphore lease")
+	}
+
+	// key is not modified, so allocate it once
+	key := backend.Key(semaphoresPrefix, req.SemaphoreKind, req.SemaphoreName)
+
+	for i := int64(0); i < acquireAttempts; i++ {
```

Comment:
I'm using `int64` here because `int64` is the base type of `time.Duration`.  Technically, it isn't necessary in this case, but I don't generally cast between numeric types unless I have some compelling reason to do so.